pos_accounting/src/pool/tests/simulation_tests.rs

```diff
@@ -0,0 +1,210 @@
+            .iter()
+            .nth(rng.gen_range(0..all_delegation_balances.len()))
+            .unwrap();
+        Some((*delegation_id, *balance))
+    } else {
+        None
+    }
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn simulation_test_db(#[case] seed: Seed) {
+    let mut rng = make_seedable_rng(seed);
+    let mut storage = InMemoryPoSAccounting::new();
+    let mut undos = Vec::<PoSAccountingUndo>::new();
+
+    let iterations_count = rng.gen_range(100..1000);
+
+    for _ in 0..iterations_count {
+        // collecting following random items every time is not efficient, because only single one of them might be used;
+        // but in this way the borrows checker can be appeased because `PoSAccountingDB` requires `&mut storage`
+        let random_pool = get_random_pool_id(&mut rng, &storage);
+        let random_delegation = get_random_delegation_data(&mut rng, &storage);
+        let random_delegation_balance = get_random_delegation_balance(&mut rng, &storage);
+
+        let mut db = PoSAccountingDB::new(&mut storage);
+        perform_random_operation(
+            &mut rng,
+            &mut db,
+            &mut undos,
+            random_pool,
+            random_delegation,
+            random_delegation_balance,
+        );
+        storage.check_consistancy();
+    }
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn simulation_test_delta(#[case] seed: Seed) {
+    let mut rng = make_seedable_rng(seed);
+    let mut storage = InMemoryPoSAccounting::new();
+    let mut undos = Vec::<PoSAccountingUndo>::new();
+
+    let iterations_count = rng.gen_range(100..1000);
+
+    for _ in 0..iterations_count {
+        // collecting following random items every time is not efficient, because only single one of them might be used;
+        // but in this way the borrows checker can be appeased because `PoSAccountingDB` requires `&mut storage`
+        let random_pool = get_random_pool_id(&mut rng, &storage);
+        let random_delegation = get_random_delegation_data(&mut rng, &storage);
+        let random_delegation_balance = get_random_delegation_balance(&mut rng, &storage);
+
+        let mut db = PoSAccountingDB::new(&mut storage);
+        let mut delta = PoSAccountingDelta::new(&db);
+        perform_random_operation(
+            &mut rng,
+            &mut delta,
+            &mut undos,
+            random_pool,
+            random_delegation,
+            random_delegation_balance,
+        );
+        db.batch_write_delta(delta.consume()).unwrap();
+        storage.check_consistancy();
+    }
+}
+
+fn perform_random_operation(
+    rng: &mut (impl Rng + CryptoRng),
+    op: &mut (impl PoSAccountingOperations + PoSAccountingView),
+    undos: &mut Vec<PoSAccountingUndo>,
+    random_pool: Option<PoolId>,
+    random_delegation: Option<(DelegationId, DelegationData)>,
+    random_delegation_balance: Option<(DelegationId, Amount)>,
+) {
+    match rng.gen_range(0..11) {
```

Comment:
One of my concerns that I've been thinking about is, what if we, in the future, add new operations? How are we gonna remember to add them to the tests?

Here's a solution: There's a crate named `variant_count`. With it, you can do:
```
assert_eq!(PoSAccountingUndo::VARIANT_COUNT, 6);
```

This way, if new ops are added, this assert will fire. With this assert, you can write a comment/message explaining what's going on and why.

Reply:
This match can be done in a much nicer way... using proper arms and probabilities... but I'm not gonna fixate on that now.